Managers/EnemyManager.gd

```diff
@@ -2,7 +2,119 @@ extends Node
+## The intervals defined by the values are continuous, successive and non-overlapping.
+@export var enemy_group_distribution: Dictionary = {
+	[1, 0, 20]: [], # an array of packed scene for enemy groups
+	[1, 20, 40]: [],
+	[1, 40, 60]: [],
+	[1, 60, 80]: [],
+	[1, 80, 100]: [],
+	[2, 0, 20]: [], # section 2
+	[2, 20, 40]: [],
+	[2, 40, 60]: [],
+	[2, 60, 80]: [],
+	[2, 80, 100]: [],
+	[3, 0, 20]: [], # section 3
+	[3, 20, 40]: [],
+	[3, 40, 60]: [],
+	[3, 60, 80]: [],
+	[3, 80, 100]: [],
+}
+
+## The group of enemies currently being used, set at the start of the combat [br]
+var current_enemy_group: EnemyGroup
+
+## The packed scene corresponding to the group of enemies, this is used to keep track of it between saves [br]
+## DO NOT FORGET to set this back to null once the fight is done [br]
+## The reset is handled via signals to be sure to not forget it when changing the battler
+var current_enemy_group_packed: PackedScene = null
+
+## The last recorded lower bound of the current sub-section
+var current_lower_bound: int = enemy_group_distribution.keys()[0][1]
+## The last recorded higher bound of the current sub-section
+var current_higher_bound: int = enemy_group_distribution.keys()[0][2]
+## The last used array of enemy group
+var enemy_group_array: Array[PackedScene]
+## The shuffled version of the `enemy_group_array`
+var shuffled_enemy_group_array: Array[PackedScene] = []
+
+func _ready() -> void:
+	if enemy_group_array.is_empty():
+		enemy_group_array.assign(enemy_group_distribution.values()[0])
+	PhaseManager.on_event_win.connect(_clear_data)
+	PhaseManager.on_defeat.connect(_clear_data)
+
+
+## This calls the inner function that actually chooses the enemy group [br]
+## This is done to save the packed scene in this EnemyManger before it is returned
+func choose_enemy_group() -> PackedScene:
+	var packed_enemy_group: PackedScene = _choose_enemy_group_inner()
+	current_enemy_group_packed = packed_enemy_group
+	return packed_enemy_group
+
+## Chooses an enemy group in the following way: [br]
+## - Check if the player is in the same sub-section as the previous time this function was called [br]
+## - Choose an array of enemy group depending on player position if either sub-section changed, or the array to choose from is empty [br]
+## - Shuffle it, pop the last element of that array
+func _choose_enemy_group_inner() -> PackedScene:
+	if current_enemy_group != null:
+		return current_enemy_group_packed
+	if DebugVar.DEBUG_USE_TEST_ENEMY_GROUP:
+		return load("res://Entity/Enemy/EnemyGroup/test_group.tscn") # return a test group
+	var height_percent_position: float = MapManager.get_map_percent_with_player_position()
+	if (current_lower_bound <= height_percent_position && current_higher_bound > height_percent_position):
+		# meaning we stayed in the same sub-section of the map, keep the same array of enemies
+		if shuffled_enemy_group_array.is_empty():
+			shuffled_enemy_group_array = enemy_group_array
```

Comment:
I believe this does not work as intended -> shuffled_enemy_group_array = enemy_group_array probably causes it to still refer to the same array, which means pop_back will remove the element from 'both'. 

Every time we reach this point, the enemy_group_array is reduced in size and eventually reaches 0 resulting in a game crash (as we return null from this function and try to instantiate it)

Reply:
Running game with breakpoint in enemy manager, enemy group array sizes:
![image](https://github.com/user-attachments/assets/10cc14ad-6591-4c06-9802-864409588455)
First combat encounter ^
![image](https://github.com/user-attachments/assets/940b961a-f7b8-485e-b444-01ab02c94ab7)
Second combat encounter ^
![image](https://github.com/user-attachments/assets/a6bbede9-d950-49d9-aa30-9253de1e5978)
Third combat encounter ^
Now the enemy group is array of size 0 and function returns null, resulting in crash

Reply:
You are totally correct about this,, I often forget that array are passed by reference, solved this in dad9ad285d4c